src/coreclr/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs

```diff
@@ -532,6 +544,142 @@ public static int GetCurrentProcessorId()
+        //   is in start-up mode. That would just add to the "rush hour" traffic.
+        private static int sCalibrationToDo;
+        private static int sCalibrationDone;
+        // 25 is chosen to budget the sampling under 100 msec total, assuming 4 msec per sample.
+        private const int CalibrationSamples = 25;
+        private static double[] sCalibrationSamples = new double[CalibrationSamples * 2];
+
+        private static void CalibrateOnce()
+        {
+            if (sCalibrationToDo >= CalibrationSamples)
+                return;
+
+            int sample = Interlocked.Increment(ref sCalibrationToDo) - 1;
+            if (sample >= CalibrationSamples)
+                return;
+
+            // Actual calibration step. Let's try to fit into ~4 msec.
+            double[] calibrationState = sCalibrationSamples;
+
+            int id = 0;
+            long t1 = 0;
+            long _1msec = Stopwatch.Frequency / 1000;
+            int iters = 1;
+
+            // double the sample size until it is 1 msec.
+            // we may spend up to 3 msec in this loop in a worst case.
+            while (t1 < _1msec)
+            {
+                iters *= 2;
+                t1 = Stopwatch.GetTimestamp();
+                for (int i = 0; i < iters; i++)
+                {
+                    id = GetCurrentProcessorNumber();
+                }
+                t1 = Stopwatch.GetTimestamp() - t1;
+            }
+
+            // assuming TLS cannot be a lot slower than ID, this should take 1 msec
+            long t2 = Stopwatch.GetTimestamp();
+            for (int i = 0; i < iters; i++)
+            {
+                UninlinedThreadStatic();
+            }
+            long t3 = Stopwatch.GetTimestamp();
+
+            // if we have useful measurements, record a sample
+            if (id >= 0 && t1 > 0 && t3 - t2 > 0)
+            {
+                calibrationState[sample * 2] = (double)t1 / iters;            // ID
+                calibrationState[sample * 2 + 1] = (double)(t3 - t2) / iters; // TLS
+            }
+            else
+            {
+                // API is not functional or clock did not go forward.
+                // just pretend it was a very expensive sample with default ratio.
+                calibrationState[sample * 2] = (double)Stopwatch.Frequency * 50; // 50 sec;
+                calibrationState[sample * 2 + 1] = Stopwatch.Frequency; // 1 sec
+            }
+
+            // If this was the last sample computed, get best times and update the ratio of ID to TLS.
+            if (Interlocked.Increment(ref sCalibrationDone) == CalibrationSamples)
+            {
+                double idMin = double.MaxValue;
+                double tlsMin = double.MaxValue;
+                for (int i = 0; i < CalibrationSamples; i++)
+                {
+                    idMin = Math.Min(idMin, calibrationState[i * 2]);       //ID
```

Comment:
Would it make sense to update the calibration as soon as we get the first sample? And have exponential backoff to the calibration intervals?

Reply:
I am concerned about two situations:
1) the first sample is very skewed and we make a bad decision that impedes the rest of measurements. For example  ProcID takes 1 second resulting in huge refresh period. It will take long time to take another sample.
2) we actually get good samples and that results in small refresh (say 5). That will make subsequent samples happen more frequently and earlier, possibly interfering with app startup. 

It does not look like we lose a lot by delaying. If app uses CoreID much, it should get it calibrated soon enough. If it does not use it much, then it cares less about precision. 

Also once we are here, it is likely that CoreID API is slow-ish and the result will not be a dramatic improvement from default.

---
In reply to: [352959076](https://github.com/dotnet/runtime/pull/467#discussion_r352959076) [](ancestors = 352959076)

Reply:
Now we do only 10 calibration cycles, this is a really small array.
I think array is worth it - since we do not rely on a single sample for anything, the resulting behavior is more predictable and the cost is not high.